### PR TITLE
Feat: GFXBuffer supports exposing staging buffer address to reduce me…

### PR DESCRIPTION
**Pull request by bluesky013**
…mcpy.

Re: # 

### Changelog

* GFXBuffer supports exposing staging buffer address to reduce memcpy.

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by bluesky013** [comment]
@cocos-robot run test cases

File: native/tests/unit-test/src/gfx_buffer_test.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    static cc::gfx::Device *device;
+
+    static void SetUpTestSuite()
+    {
+        device = cc::gfx::DeviceManager::create();
+    }
+
+    static void TearDownTestSuite()
+    {
+        CC_SAFE_DESTROY_AND_DELETE(device);
```
Comment (star-e) [inline native/tests/unit-test/src/gfx_buffer_test.cpp]: We might use unique_ptr, destroy and reset the pointer here.

File: native/cocos/renderer/gfx-base/GFXBuffer.h [hunk 2]
```diff
@@ -68,6 +93,7 @@ class CC_DLL Buffer : public GFXObject, public RefCounted {
     uint32_t _size = 0U;
     uint32_t _offset = 0U;
     BufferFlags _flags = BufferFlagBit::NONE;
+    uint8_t *_data{nullptr};
```
Comment (star-e) [inline native/cocos/renderer/gfx-base/GFXBuffer.h]: We can put _data after _isBufferView. This change might reduce size of one pointer in 64bit platform.

File: native/cocos/renderer/gfx-agent/BufferAgent.cpp [hunk 3]
```diff
@@ -54,7 +54,7 @@ BufferAgent::~BufferAgent() {
 
 void BufferAgent::doInit(const BufferInfo &info) {
     uint32_t size = getSize();
-    if (size > STAGING_BUFFER_THRESHOLD && hasFlag(_memUsage, MemoryUsageBit::HOST)) {
+    if (hasFlag(info.flags, BufferFlagBit::ENABLE_STAGING_WRITE) || (size > STAGING_BUFFER_THRESHOLD && hasFlag(_memUsage, MemoryUsageBit::HOST))) {
         for (size_t i = 0; i < DeviceAgent::MAX_FRAME_INDEX; ++i) {
             _stagingBuffers.push_back(reinterpret_cast<uint8_t *>(malloc(size)));
```
Comment (star-e) [inline native/cocos/renderer/gfx-agent/BufferAgent.cpp]: We might use std::unique_ptr<uint8_t[]> and remove usage of malloc/free.

File: native/cocos/renderer/gfx-base/GFXBuffer.h [hunk 4]
```diff
@@ -43,10 +43,34 @@ class CC_DLL Buffer : public GFXObject, public RefCounted {
...
+    template <typename T>
+    void write(const T& value, uint32_t offset) const {
+        write(reinterpret_cast<const uint8_t*>(&value), offset, sizeof(T));
+    }
+
+    void write(const uint8_t *value, uint32_t offset, uint32_t size) const;
+
     virtual void update(const void *buffer, uint32_t size) = 0;
 
+    virtual void flush(const void *buffer) { update(buffer, _size); }
```
Comment (star-e) [inline native/cocos/renderer/gfx-base/GFXBuffer.h]: Where is flush called? I didn't find the calling place in this pr.

Comment (bluesky013) [reply native/cocos/renderer/gfx-base/GFXBuffer.h:68]: I forgot to remove the interface. No need to call flush. I will delete it.

Comment (bluesky013) [reply native/cocos/renderer/gfx-base/GFXBuffer.h]: `flush` should be called in `Buffer::update`

Comment (bluesky013) [reply native/cocos/renderer/gfx-base/GFXBuffer.h]: `Buffer::flush` and `Buffer::getStagingAddress` is protected now.

File: native/cocos/renderer/gfx-agent/BufferAgent.cpp [hunk 5]
```diff
@@ -177,5 +160,13 @@ void BufferAgent::getActorBuffer(const BufferAgent *buffer, MessageQueue *mq, ui
     }
 }
 
+uint8_t *BufferAgent::getStagingAddress() const {
+    if (!_stagingBuffer) {
+        return nullptr;
+    }
+    uint32_t frameIndex = DeviceAgent::getInstance()->getCurrentIndex();
+    return _stagingBuffer.get() + _size * frameIndex;
```
Comment (minggo) [inline native/cocos/renderer/gfx-agent/BufferAgent.cpp:168]: If ENABLE_STAGING_WRITE is not set, should also return the address?

Comment (bluesky013) [reply native/cocos/renderer/gfx-agent/BufferAgent.cpp:168]: If ENABLE_STAGING_WRITE  is not set, `Buffer::write` should be disabled. `getStagingAddress` Is an internal interface. I think I should check the flag in `Buffer::write` and `Buffer::update()`.

Comment (minggo) [reply native/cocos/renderer/gfx-agent/BufferAgent.cpp:168]: Ok.

**Review by minggo — APPROVED** [review]

**Comment by bluesky013** [comment]
@cocos-robot run test cases

**Review by star-e — APPROVED** [review]

**Pull request merged**